### ISSUE TITLE
New scoping system

### ISSUE THREAD

**Issue by Ericson2314** [haskell-beam/beam#719]
> `QBaseScope` et al were put in place because of limitations of the Haskell type system at the time. It's likely we can get rid of this. In fact, we should probably rethink the entire scoping system in light of new GHC features, and be willing to break compatibility.

_Originally posted by @tathougies in https://github.com/haskell-beam/beam/issues/652#issuecomment-1439431041_

What would this look like? More data kinds? Something more radical?

**Comment by tathougies** [comment]
Oh gosh I'd have to really go back and think about this. I believe the main issue is that we ought to be able to keep the 'base' scope as an existential, but for whatever reason, when I first wrote beam, the compiler was unable to match existentials and allow typechecking to pass. It probably does at this point.

**Comment by Ericson2314** [comment]
Thanks for remembering, @tathougies.

On a similar note, here and a few other places in beam I would have used data kinds (don't need arbitrary types in our scopes do we?). Do you have thoughts on that?

**Comment by tathougies** [comment]
Yeah, the scoping system could definitely be replaced with data kinds. Keep in mind Haskell looked a lot different when I first wrote beam :)

**Comment by Ericson2314** [comment]
@tathougies I believe it, but I also think I don't know just how far back the prehistory of beam goes :).
